accessible/toolbox_treeview.component.js

```diff
@@ -0,0 +1,188 @@
+              </li>
+              <li #sendToSelected id="{{treeService.createId(sendToSelected)}}" role="treeitem" aria-selected=false [attr.aria-level]="level+2">
+                <button #sendToSelectedButton id="{{treeService.createId(sendToSelectedButton)}}" (click)="copyToMarked(block)" disabled="{{markedBlockCompatibilityHTMLText(block)}}" [attr.aria-disabled]="markedBlockCompatibilityHTMLText(block)">copy to marked spot</button>
+                {{setLabelledBy(sendToSelected, concatStringWithSpaces(sendToSelectedButton.id, "blockly-button", markedBlockCompatibilityHTMLText(block)))}}
+              </li>
+            </ol>
+            {{addClass(listItem, "blocklyHasChildren")}}
+          </li>
+          <div *ngFor="#inputBlock of block.inputList; #i=index">
+            <field-view [attr.aria-level]="level+1" *ngFor="#field of getInfo(inputBlock); #j=index" [field]="field" [level]="level+1"></field-view>
+            <toolbox-tree-view *ngIf="inputBlock.connection && inputBlock.connection.targetBlock()" [block]="inputBlock.connection.targetBlock()" [displayBlockMenu]="false" [level]="level+1"></toolbox-tree-view>
+            <li aria-selected=false #listItem1 role="treeitem" [attr.aria-level]="level+1" id="{{treeService.createId(listItem1)}}" *ngIf="inputBlock.connection && !inputBlock.connection.targetBlock()">
+              <label #label id="{{treeService.createId(label)}}">{{getInputTypeLabel(inputBlock.connection)}} {{valueOrStatement(inputBlock)}} needed:</label>
+              {{setLabelledBy(listItem1, concatStringWithSpaces("blockly-argument-text", label.id))}}
+            </li>
+          </div>
+        </ol>
+      </li>
+      <toolbox-tree-view *ngIf= "block.nextConnection && block.nextConnection.targetBlock()" [level]="level" [block]="block.nextConnection.targetBlock()" [displayBlockMenu]="false"></toolbox-tree-view>
+    `,
+    directives: [ng.core.forwardRef(
+        function() { return blocklyApp.ToolboxTreeView; }),
+        blocklyApp.FieldView],
+    inputs: ['block', 'displayBlockMenu', 'level', 'index', 'tree'],
+  })
+  .Class({
+    constructor: [blocklyApp.ClipboardService, blocklyApp.TreeService,
+                  function(_clipboardService, _treeService) {
+      this.infoBlocks = {};
+      this.nextBlock = {};
+      this.sharedClipboardService = _clipboardService;
+      this.treeService = _treeService;
+    }],
+    setLabelledBy: function(item, id) {
+      if (!item.getAttribute('aria-labelledby')) {
+        item.setAttribute('aria-labelledby', id);
+      }
+    },
+    concatStringWithSpaces: function() {
+      var string = arguments[0];
+      for (i = 1; i < arguments.length; i++) {
+        var arg = arguments[i];
+        // arg can be undefined and if so, should be ignored.
+        if (arg) {
+          string += ' ' + arg;
+        }
+      }
+      return string;
+    },
+    createCategoryDependantId: function(index, parentList) {
```

Comment:
Actually, this looks like it has a bunch of side-effects. Can we avoid that, and perhaps have this function be getId() instead? If other things need to happen, use a separate function.